### PR TITLE
Permit layer averaging in 3D (default to layers of extruded mesh)

### PR DESCRIPTION
**Pull request by drhodrid**
This branch makes a few changes to the layer average routine, implemented by @stephankramer. For extruded meshes it defaults to the layers of the extruded mesh. It also permits use of our layer average routines in 3D.

I have updated our 3-D spherical demo to use the new averager. I have added a diagnostic of the average temperature deviation across the domain. @angus-g - could you add the latter to our check values?

### CODE REVIEW

File: demos/3d_spherical/3d_spherical.py [hunk 1]
```diff
@@ -125,10 +115,10 @@ def layer_average(T):
 
     # Write output:
     if timestep % dump_period == 0:
-        # compute radial temperature
-        Tavg = layer_average(T)
+        # compute radially averaged temperature profile
+        averager.extrapolate_layer_average(T_avg, averager.get_layer_average(T))
         # compute deviation from layer average
-        T_dev.assign(T-Tavg)
+        T_dev.project(T-T_avg)
```
Comment (stephankramer) [inline demos/3d_spherical/3d_spherical.py:121]: This is probably due to the issue we discussed, but this can now be an assign again, right?

**Comment by drhodrid** [comment]
I think so, yes - although I got a memory issue on Gadi when I tested that. Running a case now to confirm.

----------------------------------------------------------------------------------------------------------
Professor (Computational & Observational Geodynamics)
Associate Director Research & Engagement
Research School of Earth Sciences,
The Australian National University,
Mills Road, Acton,
Canberra, ACT, 2601, Australia

T: +61 (0)2 6125 3643
E: ***@***.***
W: https://earthsciences.anu.edu.au/people/academics/professor-rhodri-davies
CRICOS Provider No. 00120C
----------------------------------------------------------------------------------------------------------

On 16 Oct 2023, at 7:43 pm, Stephan Kramer ***@***.***> wrote:



@stephankramer commented on this pull request.

________________________________

In demos/3d_spherical/3d_spherical.py<https://github.com/g-adopt/g-adopt/pull/37#discussion_r1360317156>:

>          # compute deviation from layer average
-        T_dev.assign(T-Tavg)
+        T_dev.project(T-T_avg)


This is probably due to the issue we discussed, but this can now be an assign again, right?

**Comment by angus-g** [comment]
Looks like maybe UFL doesn't like the negative indexing on `SpatialCoordinate`?

```
  File "/__w/g-adopt/g-adopt/demos/adjoint/forward.py", line 50, in <module>
    averager = LayerAveraging(
  File "/home/firedrake/firedrake/lib/python3.10/site-packages/gadopt/utility.py", line 353, in __init__
    self.r = XYZ[-1]
  File "/home/firedrake/firedrake/src/ufl/ufl/exproperators.py", line 395, in _getitem
    a = Indexed(self, mi)
  File "/home/firedrake/firedrake/src/ufl/ufl/indexed.py", line 72, in __init__
    raise ValueError("Fixed index out of range!")
```

**Review by sghelichkhani — APPROVED** [review]
Seems all good to go! Sorry about the confusion earlier!

**Pull request merged**